accepted/0000-link-packages-to-source-and-build.md

```diff
@@ -0,0 +1,632 @@
+| Ref               | GITHUB_CONTEXT.ref         | ??               | ??       |
+| Git SHA           | GITHUB_CONTEXT.sha         | ??               | ??       |
+| OS image          | ImageOS                    | ??               | ??       |
+| OS image version  | ImageVersion               | ??               | ??       |
+| Host architecture | RUNNER_ARCH                | ??               | ??       |
+
+#### Signature envelope
+The signed build provenance attestation must be stored and transferred together with its signature. The envelope carries this functionality. We plan to support [DSSE envelopes](https://github.com/secure-systems-lab/dsse), and preferably [COSE envelopes](https://datatracker.ietf.org/doc/html/rfc8152) as well.
+
+For a more detailed look into different envelope schemas, see this [comparison of signing envelope formats](https://docs.google.com/document/d/18YVGA4mq45wfUkWrAqWkymzdHRcXxlwINKXnEp86L0w/edit#heading=h.3v6vw85pjg5e).
+
+### Release attestations
+When a package is successfully published, the npm registry will "attest" the release and upload this to Rekor. This includes a signature that links the repository to the package name. This must be verified by the CLI when during verification.
+
+This can be achieved by publishing the package signature to Rekor by using the proposed [“package entry”](https://github.com/sigstore/rekor/pull/805) log type.
+
+The release attestation will contain the following claims, linking the build to the release:
+- Package org/name
+- Package version
+- Registry signature
+- Tarball shasum
+
+### Logging additional events on the Rekor transparency log
+
+The current proposal only suggests publishing two events to the public Rekor log: sign during build and release/publish.
+
+Publishing a richer set of events to the log could help create a safe and independent history of sensitive events. This could provide the possibility of detection when the registry DB or blob bucket are compromised, because it would be out of sync with the transparency log.
+
+There's interest from OpenSSF's Securing Software Repos WG to standardize on a set of events to easy interoperability between different package registries.
+
+Possible events that could be recorded in the transparency log:
+
+- User registration
+- Sign during build
+- Push
+- Publish or release
+- Yank or delete
+- Owner added
+- Owner removed
+- Owner changed email address
+- Owner enabled MFA
+- Owner disabled MFA
+- Endorsement by non-owner
+- Ownership call
+- Ownership request
+- Ownership transfer
+- Resurrection of name
+
+## Prior Art
+### Sigstore cosign CLI
+The [cosign](https://github.com/sigstore/cosign) CLI tool supports signing arbitrary blobs and could be used to sign the package tarball manually but requires significant setup from maintainers and offers no simple way for consumers to verify signatures unless you know you to query for the package signature on Rekor.
```

Comment:
This is also less safe if we allow non-trusted builds to sign the tarball as the build could falsify some elements of its provenance more easily.

Reply:
Yes! Using a CI system to get the non-falsifiable context (e.g. git ref/sha) is a very important step.

Reply:
How is that not falsifiable? CI systems can run `git` commands and set the sha to whatever they want.

Reply:
> How is that not falsifiable? CI systems can run `git` commands and set the sha to whatever they want.

This is a good point. As I see it, there are two sides to provenance information. One part we can get from the CI/CD systems OIDC id token. The content of this can be verified against the CIs public key so can't be falsified by some code running in CI/CD unless the CI provide is compromised. 

The id token will tell us which repo the action or CI job was triggered in but not necessarily which repo the build actually ran in. The information in the id token allows us to link back to the build logs for example (unless of course the repo has been deleted, which is still a problem here).

The other side is which repo the build was actually running in. This is slightly harder to lock down. This is why we're proposing a "trusted builder" that's a reusable workflow maintained by npm that can be used to build and publish the package. Using this workflow is recommended but optional as it might not be feasible to support every kind of publish workflow.

The trusted builder will execute the steps to checkout the repo from the information in the ID token. There's still a risk that the user supplied build steps run some code that modifies the checked out source repo. Locking this down further will be very hard until we have a way to run certain steps, e.g. `npm run build` without network access.

Reply:
Maybe a more complete way of wording it would be end user non-falsifiable. Assuming you trust GitHub to do the right things then they should be the ones using `git` on your behalf. Compare that with the actual end user builds which can just run any arbitrary set of things.

Reply:
It's also worth noting that the non-forgable part comes from the GitHub workflow run and OIDC identity attached to it. The idea is that a random person could not forge this identity but running the same commands on their laptop.